opm/autodiff/StandardWell_impl.hpp

```diff
@@ -1318,6 +1292,188 @@ namespace Opm
+
+        // checking the BHP limit related
+        checkOperabilityUnderBHPLimitProducer(ebos_simulator);
+
+        // TODO: if the BHP limit does not work anyway, we do not need to do the following
+        // We do it now for studying purpose.
+        // checking whether the well can operate under the THP constraints.
+        if (this->wellHasTHPConstraints()) {
+            checkOperabilityUnderTHPLimitProducer(ebos_simulator);
+        }
+
+        // checking whether the well can not produce/inject
+        this->operability_status_.existing_drawdown_correct_direction =
+                                                   ! allDrawDownWrongDirection(ebos_simulator);
+
+        const bool well_operable = this->operability_status_.isOperable();
+
+        if (!well_operable && old_well_operable) {
+            OpmLog::info(" well " + name() + " gets SHUT during iteration ");
+        } else if (well_operable && !old_well_operable) {
+            OpmLog::info(" well " + name() + " gets REVIVED during iteration ");
+        }
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    void
+    StandardWell<TypeTag>::
+    checkOperabilityUnderBHPLimitProducer(const Simulator& ebos_simulator)
+    {
+        const double bhp_limit = mostStrictBhpFromBhpLimits();
+        // TODO: a better way to detect whether the BHP is defaulted or not
+        if ( bhp_limit > 1.5e5 || !this->wellHasTHPConstraints() ) {
+            // if the BHP limit is not defaulted or the well does not have a THP limit
+            // we need to check the BHP limit
+
+            for (int p = 0; p < number_of_phases_; ++p) {
+                const double temp = ipr_a_[p] - ipr_b_[p] * bhp_limit;
+                if (temp < 0.) {
+                    this->operability_status_.operable_under_only_bhp_limit = false;
+                    break;
+                }
+            }
+
+            // checking whether running under BHP limit will violate THP limit
+            if (this->operability_status_.operable_under_only_bhp_limit && this->wellHasTHPConstraints()) {
+                // option 1: calculate well rates based on the BHP limit.
+                // option 2: stick with the above IPR curve
+                // we use IPR here
+                const double bhp_limit = mostStrictBhpFromBhpLimits();
+                std::vector<double> well_rates_bhp_limit;
+                computeWellRatesWithBhp(ebos_simulator, bhp_limit, well_rates_bhp_limit);
+
+                const double thp = calculateThpFromBhp(well_rates_bhp_limit, bhp_limit);
+                const double thp_limit = this->getTHPConstraint();
+
+                if (thp < thp_limit) {
+                    this->operability_status_.obey_thp_limit_under_bhp_limit = false;
+                }
+            }
+        } else {
+            // defaulted BHP and there is THP constraints
+            // default BHP limit is about 1.03 bar.
+            // when applied the hydrostatic pressure correction,
```

Comment:
I do not understand this comment. Why can we get a negative bhp value? Here it is assumed equal to 1 atm. Did you mean negative thp value?

Reply:
The bhp we calculate from the VFP table is (bhp + dp). With a so low bhp, dp is negative (typically with larger value), (bhp + dp) will be negative unless bhp reference depth is close to vfp reference depth.